core/src/main/java/com/sequenceiq/cloudbreak/reactor/handler/cluster/StopStartDownscaleDecommissionViaCMHandler.java

```diff
@@ -77,22 +83,40 @@ protected Selectable doAccept(HandlerEvent<StopStartDownscaleDecommissionViaCMRe
 
         try {
             Stack stack = stackService.getByIdWithLists(request.getResourceId());
-            Cluster cluster = stack.getCluster();
             ClusterDecomissionService clusterDecomissionService = clusterApiConnectors.getConnector(stack).clusterDecomissionService();
 
             Set<String> hostNames = getHostNamesForPrivateIds(request.getInstanceIdsToDecommission(), stack);
             LOGGER.debug("Attempting to decommission hosts. count={}, hostnames={}", hostNames.size(), hostNames);
 
             Map<String, InstanceMetadataView> hostsToRemove = clusterDecomissionService.collectHostsToRemove(request.getHostGroupName(), hostNames);
+            Map<String, InstanceMetadataView> finalHostsToRemove = hostsToRemove;
             List<String> missingHostsInCm = Collections.emptyList();
             if (hostNames.size() != hostsToRemove.size()) {
                 missingHostsInCm = hostNames.stream()
-                        .filter(h -> !hostsToRemove.containsKey(h))
+                        .filter(h -> !finalHostsToRemove.containsKey(h))
                         .collect(Collectors.toList());
                 LOGGER.info("Found fewer instances in CM to decommission, as compared to initial ask. foundCount={}, initialCount={}, missingHostsInCm={}",
                         hostsToRemove.size(), hostNames.size(), missingHostsInCm);
             }
 
+            List<InstanceMetadataView> instancesWithServicesNotRunning = instanceIdToInstanceMetaDataConverter
+                    .getNotDeletedAndNotZombieInstances(instanceMetaDataService.getAllAvailableInstanceMetadataViewsByStackId(stack.getId()),
+                            request.getHostGroupName(),
+                            request.getRunningInstancesWithServicesNotRunning().stream().map(CloudInstance::getInstanceId).collect(Collectors.toSet()));
+
+            List<InstanceMetadataView> additionalInstancesToDecommission = instancesWithServicesNotRunning
+                    .stream().filter(im -> !finalHostsToRemove.containsKey(im.getDiscoveryFQDN())).collect(Collectors.toList());
+
+            if (!additionalInstancesToDecommission.isEmpty()) {
+                LOGGER.info("Including instancesWithServicesNotRunning as part of decommission node list");
+                Collection<InstanceMetadataView> finalInstancesToDecommission = hostsToRemove.values();
+                finalInstancesToDecommission.addAll(additionalInstancesToDecommission);
+                hostsToRemove = finalInstancesToDecommission.stream().collect(Collectors.toMap(
+                        InstanceMetadataView::getDiscoveryFQDN,
+                        Function.identity()
+                ));
+            }
+
```

Comment:
The CM decommission API.
Does this API work on hosts which are 1) In maintenence mode, or 2) Already decommissioned, or 3) BAD_HEALTH
The recoveryCandidates will likely need to be separated. Example: Nodes DECOMMISSIONED + ALREADY_IN_MAINT_MODE - don't need to be part of this list.
Completely depends on how the API behaves.

One of the more likely scenarios is that decommission succeeded, but STOP failed. Should such instances (DECOMMISSIONED, HEALTHY, SERVICES_NOT_RUNNING, In_maint_mode) - be a passthrough in this step?, and go straight to the next step to be STOPPED.

Reply:
Also, is there some way we can protect a regular downscale from being affected by a failure because of the recovery candidates?
Maybe retry without the recovery candidates (can get slow and repetitive, so not great) - and depends on CM API since the first call could have put some nodes into a decommission state. 

Maybe something more bizarre like a failure os a result of attempted recovery implies - the next scale operation is only recovery, or completely skips recovery)

I think the recovery collection needs to be protected by ENTITLEMENT - at least initially (and easy way to turn it off, and convert into empty lists)

Reply:
Also does the maint mode API work on hosts which are already in maintenance mode ?

Reply:
refactored the recovery candidate collection as discussed. Also tested CM APIs with hosts already present in target states, and the APIs are still working as expected